tool/add_missing_c.rb

```diff
@@ -0,0 +1,96 @@
+# Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved.
+# This code is released under a tri EPL/GPL/LGPL license. You can use it,
+# redistribute it and/or modify it under the terms of the:
+#
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+
+# This tool is designed to search the truffleruby C headers for function definitions which
+# have not been defined in the ruby.c file. If a definition is missing, it will print the 
+# corresponding C definition found in the MRI sources with an error message as the body of
+# the function.
+
+# This is assumed to be run from the truffleruby project directory 
+# with the MRI project as a sibling "../ruby"
+# ruby tool/add_missing_c.rb
+
+header_files = Dir.glob("lib/cext/**/*.h")
+header_files.reject! { |h| h.include?("truffle") }
+ruby_path = File.expand_path("../ruby")
+ruby_c_files = Dir.glob("#{ruby_path}/**/*.c")
+
+header_definitions = []
+excludes = [
+            "rb_ia64_flushrs",            # Not found in MRI c
+            "rb_enc_casefold",            # Not found in MRI c
+            "rb_interned_id_p",           # Not found in MRI c
+            "rb_sym_interned_p",          # Not found in MRI c
+            "rb_random_int",              # Not found in MRI c
+            "rb_trap_exec",               # Not found in MRI c
+            "rb_str_buf_new2",            # Defined by macro definition
+            "rb_str_buf_cat2",            # Defined by macro definition
+            "rb_str_dup_frozen",          # Defined by macro definition
+            "rb_hash_uint32",             # Defined by macro definition
+            "rb_hash_uint",               # Defined by macro definition
+            "rb_hash_end",                # Defined by macro definition
+            "rb_check_safe_str",          # Defined by macro definition
+            "rb_w32_times",               # Skip w32 functions
+            "rb_w32_read",                # Skip w32 functions
+            "rb_w32_write",               # Skip w32 functions
+            "rb_w32_write_console",       # Skip w32 functions
+            "rb_w32_asynchronize",        # Skip w32 functions
+            "rb_w32_pow",                 # Skip w32 functions
+            "rb_scan_args",               # TruffleRuby defines this as a macro
+            "rb_funcall",                 # TruffleRuby defines this as a macro
+            "rb_enumeratorize_with_size", # Compilation error when defined
+          ]
+
+header_files.each do |f|
+    File.open(f).each_line do |line|
```

Comment:
`File.foreach` is better here, as it doesn't leak a file descriptor.

Reply:
Fixed.